File: 프로그래머스/3/43164. 여행경로/여행경로.py
# 프로그래머스 43164 여행 경로 LV.3
def solution(tickets):
    tickets.sort(key=lambda x: (x[0], x[1]))

    def dfs(t, path):
        if len(t) == 0:
            return path

        now = path[-1]
        idx = -1

        for i in range(len(t)):
            if now == t[i][0]:
                idx = i
                break

        if idx == -1:
            return []

        while t[idx][0] == now:
            next_path = dfs(t[:idx] + t[idx + 1:], path + [t[idx][1]])

            if next_path:
                return next_path

            idx += 1

    return dfs(tickets, ['ICN'])

File: 프로그래머스/3/43164. 여행경로/여행경로.java
import java.util.*;

class Solution {
    boolean[] visited;
    List<String> result = new ArrayList<>();
    
    public String[] solution(String[][] tickets) {
        visited = new boolean[tickets.length];
        dfs(tickets, "ICN", "ICN", 0);
        Collections.sort(result);
        return result.get(0).split(" ");
    }
    
    public void dfs(String[][] tickets, String start, String route, int cnt) {
        if (cnt == tickets.length) {
            result.add(route);
        } else {
            for (int i = 0; i < tickets.length; i++) {
                if (!visited[i] && tickets[i][0].equals(start)) {
                    visited[i] = true;
                    dfs(tickets, tickets[i][1], route + " " + tickets[i][1], cnt + 1);
                    visited[i] = false;
                }
            }
        }
    }
}